File: Project2/range.cs
/*
-- Developer    : Don Dang
-- Project      : Project 2
-- Instructor   : Dr. Dingle
-- File Name    : range.cs
-- File Version : 1.0
-- Due Date	    : 4/17/2019
-- Course Name  : CPSC 3200
----------------------------------------------------------------------------
OVERVIEW: 
 The Range class will accept two factor values and use them to initial two 
  different Factor objects.
 A passed-in ping number will be checked if it's a multiple of the factor
  value. If Diverized then increment ping count
 Every successful Div() called will invoke setting the min & max ping stats.

Design Decisions and Assumptions:
 The range class encapsulates its private data members.
 There is no Default Constructor, a DEFALUT CONSTRUCTOR VALUE will be provided
 Factor value can not be zero.
   If factor values = 0, then they will get assign to DEFAULT CONSTRUCTOR VALUE.
 DEFAULT CONSTRUCTOR VALUE will be postive integer greater than 1.
 Factor values passed to constructors can be duplicated.
 All public methods can be called regardless of Factor object status.
 After every called to the Factor class's Div(), all Factor object will 
  be set to active.
------------------------------------------------------------------------------
Interface Invariants 
 Ping() 
  Increment and return the ping count when the passed-in number is a multiple
    of two factor values

 GetMaxPing()
  Return the greatest successful pinged number.
  It could be a 0 if there is no successful pinged number

 GetMinPing()
  Return the lowest successful pinged number.
  It could be a 0 if there is no successful pinged number
 
 GetPingCount()
  Return a count of successful pinged number.
  It could be a 0 if there is no successful pinged number

 Replace()
  Replace the current two factor values with two new factor values

---------------------------------------------------------------------------------
IMPLEMENTATION INVARIANTS:
 Allow max two factor values can be passed into Range Constructor.
 Allow factor values to be changed after firing constructor.

 Ping() 
  Determine if the passed-in number is a multiple of two factor values.
  Validation is checked by comparing the returned value of divCount 
   for each factor object.
  If it is a successful multiple then, 
   1. Increment ping count.
   2. Update the min & max stats.
  After every called to the Factor class's Div(), the Factor object get reset. This
   means the Factor object will never be inactive.
  Return ping count.

 GetMaxPing()
  Accessor to return max successful pinged value

 GetMinPing()
   Accessor to return min successful pinged value
 
 GetPingCount()
   Accessor to return total successful pinged count

 Replace()
  Accept two factor values
  Separately used each passed-in factor value to initialize a new Factor object 
   with passed-in factor value
  Current Factor object's factor values will be replaced by new factor values
  Reset all private stats data members to initial values
----------------------------------------------------------------------------------     
CLASS INVARIANTS:
  Increment and return the ping count when the passed-in number is a multiple
    of two factor values
----------------------------------------------------------------------------------
-- WHEN		WHO		WHAT
-- 4/11/19	DD  	Added comments for Interface Invariants, 
--                    Implementation Invariants, Class Invariants
*/


using System;

namespace Project2
{
    class range
    {
        private factor factorObjOne;
        private factor factorObjTwo;
        private uint ping_Count;
        private uint min_Ping;
        private uint max_Ping;
        const int INITIAL_VALUE = 0;
        const int DEFAULT_CONSTRUCTOR_FACTOR1 = 2;
        const int DEFAULT_CONSTRUCTOR_FACTOR2 = 4;

   
        public range(uint factorValue1 = DEFAULT_CONSTRUCTOR_FACTOR1, 
            uint factorValue2 = DEFAULT_CONSTRUCTOR_FACTOR2)
        {
            if (factorValue1 == 0)
            {
                factorValue1 = DEFAULT_CONSTRUCTOR_FACTOR1;
            }

            if (factorValue2 == 0)
            {
                factorValue2 = DEFAULT_CONSTRUCTOR_FACTOR2;
            }

            factorObjOne = new factor(factorValue1);
            factorObjTwo = new factor(factorValue2);
            min_Ping = INITIAL_VALUE;
            max_Ping = INITIAL_VALUE;
            ping_Count = INITIAL_VALUE;
        }

        //PRE : N/A
        //POST: Factor object will be active
        public uint Ping(uint pNumber)
        {
             int retDivCount1 = factorObjOne.Div(pNumber);
             int retDivCount2 = factorObjTwo.Div(pNumber);
              factorObjOne.Reset();
              factorObjTwo.Reset();
              if ((retDivCount1 == 1) && (retDivCount2 == 1))
              {
                ping_Count++;
                UpdateMinMaxPingRange(pNumber);
              }
             
            return ping_Count;
        }

        //PRE : N/A
        //POST: N/A
        public uint GetMaxPing()
        {
            return max_Ping;
        }

        //PRE : N/A
        //POST: N/A
        public uint GetMinPing()
        {
            return min_Ping;
        }

        //PRE : N/A
        //POST: N/A
        public uint GetPingCount()
        {
            return ping_Count;
        }
 
        //PRE : N/A
        //POST: Factor object will be active
        public void Replace(uint factorValue1, uint factorValue2)
        {
            factorObjOne = new factor(factorValue1);
            factorObjTwo = new factor(factorValue2);
            min_Ping = INITIAL_VALUE;
            max_Ping = INITIAL_VALUE;
            ping_Count = INITIAL_VALUE;
        }

        //PRE : N/A
        //POST: N/A
        private void UpdateMinMaxPingRange(uint number)
        {
            if (max_Ping == 0 && min_Ping == 0)
            {
                max_Ping = number;
                min_Ping = number;
            }
            else
            {
                if (number > max_Ping)
                {
                    max_Ping = number; 
                }
                else
                {
                    if ((min_Ping == 0) || (number < min_Ping))
                    {
                        min_Ping = number; 
                    }
                }
            }
        }
    }
}

File: p5/range.cpp
/*
-- Developer    : Don Dang
-- Project      : Project 5
-- Instructor   : Dr. Dingle
-- File Name    : range.cpp
-- File Version : 1.1
-- Due Date	    : 5/24/2019
-- Course Name  : CPSC 3200
---------------------------------------------------------------------------------------------------------------------------------
 * IMPLEMENTATION INVARIANTS:
 *  1 - All range objects are kept active internally by calling reset after each pinged comparison.
 *
 *  2 - The factor objects are initialized either through client input, or through the default values already assigned.
 *
 *  3 - The Ping() method calls methods Div() to determine if the input value was a ping.
 * ---------------------------------------------------------------------------------------------------------------------------------
 * CLASS INVARIANTS:
 *  1 - Constructor initializes two factor objects & all other data members to keep track
 *      of statistics.
 *
 *	2 - Overloaded Operators:
 *
 *		Arithmetic operators (+, -) add and subtract from the encapsulated_factor value within the encapsulated_factor
 *		values located in the range object. Support of mixed mode with integers is implemented. In mixed mode, the integer is added or
 * 		subtracted both encapsulated_factor values in the range object.
 *
 *		Relational operators (==, !=, <, <=, >, >=) are implemented & will compare  encapsulated_factor
 *		values within the range object. Range objects can be compared to other range objects, range objects
 *		can be compared to integers, and integers can be compared to range objects. The result returned a boolean
 *		from the comparison. The comparison is only true is both encapsulated_factor values in the range object
 *		pass the relational check with what it is compared against.
 *
 *		Unary operators (++, --) increments or decrements both encapsulated_factor values of a range object by one.
 *		Pre and post increment and decrement are supported. These operators are destructive.
 *
 *		Assignment operators (+=, -=) are destructive, meaning they will alter the state of an existing factor
 *		objects encapsulated_factor rather than creating and returning a new factor object. These operators are
 *		supported for factor objects with other factor objects, factor objects with integers, and integers with
 *		factor objects. The encapsulated_factor values are what is used for addition and subtraction and += adds
 * 		a value on the RHS and assigns it to the LHS and the -= does the same but with subtraction instead. Same
 * 		as (+, -) but destructive. Supported for range and range operation
 * 		and range and integer operation.
 *
 *		Bitwise operator (<<) will output the encapsulated_factors ONLY for testing.
 * 		Example: cout << range; will return both encapsulated_factor values of object range seperated by a comma.
 *
 *		NOTES:
 *
 *		Subtraction operators (-, --, -=) the encapsulated values will not subtract a value if the
 *		resulting encapsulated factor values will be less than 1. 

-------------------------------------------------------------------------------
Range Feedback: Range Class Design - 4
	- 1 redundant boolean evualations in update min and max
	- 1 not self - documenting names
	- 2 factor values should not be allowed to be 1
----------------------------------------------------------------------------------
-- WHEN     WHO		WHAT
  5/20/19   DD      Ported range.cs to range.cpp
									Updated range.cpp from feedback (See above Range Feedback section)
*/

#include "range.h"

range::range(unsigned int factor1, unsigned int factor2) //: factObject1(factor1), factObject2(factor2)
{
	if (factor1 <= 1) factor1 = DEFAULT_FACTOR_VALUE;
	if (factor2 <= 1) factor2 = DEFAULT_FACTOR_VALUE;

	factObject1 = factor(factor1);
	factObject2 = factor(factor2);
	min_Ping = DEFAULT_PING_VALUE;
	max_Ping = DEFAULT_PING_VALUE;
	pingsCount = DEFAULT_PING_VALUE;
}

range::~range()
{
}

unsigned int range::Ping(unsigned int pNumber)
{
	bool isMultiple1 = factObject1.Div(pNumber);
	bool isMultiple2 = factObject2.Div(pNumber);
	factObject1.Reset();
	factObject2.Reset();
	if ((isMultiple1) && (isMultiple2))
	{
		pingsCount++;
		UpdateMinMaxPingValue(pNumber);
	}

	return pingsCount;
}

void range::UpdateMinMaxPingValue(unsigned int number)
{
		if (number >= max_Ping)
		{
			max_Ping = number;
		}
		else
		{
			if (number <= min_Ping)
			{
				min_Ping = number;
			}
		}
}

unsigned int range::GetMaxPingNumber() {return max_Ping;}

unsigned int range::GetMinPingNumber() {return min_Ping;}

unsigned int range::GetTotalPingCount() { return pingsCount; }

void range::Replace(unsigned int rFactor1, unsigned int rFactor2)
{
	factObject1 = factor(rFactor1);
	factObject2 = factor(rFactor2);
	min_Ping = DEFAULT_PING_VALUE;
	max_Ping = DEFAULT_PING_VALUE;
	pingsCount = DEFAULT_PING_VALUE;
}

/////////////////Stream I/O/////////////////////////
ostream& operator<<(ostream& os, const range & r)
{
	os << r.factObject1 << "," << r.factObject2;
	return os;
}


/////////////////////////////////RELATION OPERATORS/////////////////////////////////////

bool range::operator== (const range & rhs) { return factObject1 == rhs.factObject1 && factObject2 == rhs.factObject2;}
bool range::operator== (unsigned int x){ return (factObject1 == x && factObject2 == x);}
bool operator== (unsigned int x, const range& rhs) { return (x == rhs.factObject1 && x == rhs.factObject2);}

bool range::operator<(const range & rhs) { return (factObject1 < rhs.factObject1 && factObject2 < rhs.factObject2); }
bool range::operator<(unsigned int x) {return (factObject1 < x && factObject2 < x);}
bool operator<(unsigned int x, const range& rhs) { return (x < rhs.factObject1 && x < rhs.factObject2); }

bool range::operator>(const range & rhs) { return (factObject1 > rhs.factObject1 && factObject2 > rhs.factObject2); }
bool range::operator>(unsigned int x) { return (factObject1 > x && factObject2 > x); }
bool operator>(unsigned int x, const range& rhs) { return (x > rhs.factObject1 && x > rhs.factObject2); }


bool range::operator!=(const range & rhs){return (factObject1 != rhs.factObject1 && factObject2 != rhs.factObject2);}
bool range::operator!=(unsigned int x) { return (factObject1 != x && factObject2 != x); }
bool operator!=(unsigned int x, const range& rhs) { return (x != rhs.factObject1 && x != rhs.factObject2); }


bool range::operator<=(const range & rhs){return (factObject1 <= rhs.factObject1 && factObject2 <= rhs.factObject2);}
bool range::operator<=(unsigned int x) { return (factObject1 <= x && factObject2 <= x); }
bool operator<=(unsigned int x, const range& rhs) { return (x <= rhs.factObject1 && x <= rhs.factObject2); }


bool range::operator>=(const range & rhs){return (factObject1 >= rhs.factObject1 && factObject2 >= rhs.factObject2);}
bool range::operator>=(unsigned int x) { return (factObject1 >= x && factObject2 >= x); }
bool operator>=(unsigned int x, const range& rhs) { return (x >= rhs.factObject1 && x >= rhs.factObject2); }

///////////////////////////////MATHEMATICAL OPERATORS//////////////////////////////////

range range::operator+(unsigned int x)
{
	range local(*this);
	local.factObject1 += x;
	local.factObject2 += x;
	
	return local;
}

//range operator+(unsigned int x, const range & rhs)
//{
//	range local(rhs);
//	local.factObject1 += x;
//	local.factObject2 += x;
//	return local;
//}


range range::operator-(unsigned int x)
{

	range local(*this);
	local.factObject1 -= x;
	local.factObject2 -= x;
	return local;
}

//range operator-(unsigned int x, const range & rhs)
//{
//	range local(rhs);
//	local.factObject1 -= x;
//	local.factObject2 -= x;
//
//	return local;
//}

////////////////////////////////SHORTCUT OPERATORS////////////////////////////////////////////
range range::operator++()
{

	range local(*this);
	++local.factObject1;
	++local.factObject2;
	return local;
}

range range::operator++(int x)
{
	range temp(*this);
	range local(*this);
	local.factObject1++;
	local.factObject2++;
	return temp;
}

range range::operator--()
{
	range local(*this);
	--local.factObject1;
	--local.factObject2;
	return local;
}

range range::operator--(int x)
{
	range temp(*this);
	range local(*this);
	local.factObject1--;
	local.factObject2--;
	return temp;
}

/////////////////////////////////COMPOUNDED ASSIGNMENT///////////////////////////
void range::operator+=(unsigned int x)
{
	factObject1 += x;
	factObject2 += x;
}

void range::operator-=(unsigned int x)
{
	factObject1 -= x;
	factObject2 -= x;
}

